src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/FunctionProbes/ProfilerAbi.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//
+// !!IMPORTANT!!
+// All types in this file are also used by the mutating profiler during PINVOKEs and so **need** to be kept in sync
+// with the profiler's version (found in src\Profilers\MutatingMonitorProfiler\ProbeInstrumentation\ProbeInjector.h).
+//
+
+namespace Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing.FunctionProbes
+{
+    internal enum SpecialCaseBoxingTypes : uint
+    {
+        Unknown = 0,
+        Object,
+        Boolean,
+        Char,
+        SByte,
+        Byte,
+        Int16,
+        UInt16,
+        Int32,
+        UInt32,
+        Int64,
+        UInt64,
+        IntPtr,
+        UIntPtr,
+        Single,
+        Double,
+    };
+
+    internal enum InstructionType : ushort
+    {
+        Unknown = 0,
+        SpecialCaseToken,
+        MetadataToken
+    }
+
+    internal struct ParameterBoxingInstructions
```

Comment:
At the very least, you should have LayoutKind.Sequential here. It might be useful to add a comment regarding the fact that Token is a union on the native side.

Reply:
You may need to look into Pack as well to make sure packsize is correct in both managed and native.

Reply:
structs are LayoutKind.Sequential by default

Reply:
It probably should be `LayoutKind.Explicit`: https://learn.microsoft.com/en-us/dotnet/standard/native-interop/customize-struct-marshalling#customize-structure-layout

Reply:
As for controlling the alignment, structs use the platform's default alignment by default. This is the same as our profiler code so unless we're somehow loading a 32 bit profiler into a 64 bit .net app we shouldn't need to worry about hard coding alignment.

Reply:
> It probably should be `LayoutKind.Explicit`: https://learn.microsoft.com/en-us/dotnet/standard/native-interop/customize-struct-marshalling#customize-structure-layout

I don't know if I agree with that assertion. sequential makes the most sense here since we have no need to override the field offsets since no unions are used.

Reply:
From the docs: "DO only use LayoutKind.Explicit in marshalling when your native struct also has an explicit layout, such as a union."

The native struct `PARAMETER_BOXING_INSTRUCTIONS` has a union, thus it has an explicit layout, correct? Thus this rule says you should use `LayoutKind.Explicit`.

Reply:
All of the fields in the union are the same size, so explicit layout isn't needed to account for that. I'll add a static assert to ensure we don't accidentally change the union's size in the future

Reply:
Discussed offline, I'll add a comment to the token field and explicitly call out the struct layout